### PR TITLE
fix(vanilla, react): use experimental_use and some refactors

### PR DESCRIPTION
**Pull request by dai-shi**
Ref: https://github.com/facebook/react/pull/25084

In addition:
- Refactor with unified Symbol `PROXY_STATE`.
- Precise `SNAPSHOT` type with Awaited.

### CODE REVIEW

File: src/react.ts [hunk 1]
```diff
@@ -1,7 +1,17 @@
-import { useCallback, useDebugValue, useEffect, useMemo, useRef } from 'react'
 import {
-  affectedToPathList,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  experimental_use,
```
Comment (SukkaW) [inline src/react.ts:4]: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62189

It seems that the typing of `experimental_use` is added to the `@types/react`, and `ts-ignore` could be removed.

Comment (dai-shi) [reply src/react.ts:4]: Oh, yeah, I updated [there](https://github.com/jotai-labs/jotai-signal) and [there](https://github.com/jotai-labs/jotai-uncontrolled), but forgot here. Thanks.

**Comment by dai-shi** [comment]
According to https://github.com/reactjs/rfcs/pull/229, the use of `use` here is violating the rule/convention of React.
From our perspective, this PR is not breaking.
When React finalizes the spec and release a stable version, we will consider following their pattern, and in that case, our library comes with **breaking** changes. i.e. snapshot will not resolve promises.
Meanwhile, let's merge this and release a patch version.

File: src/react.ts [hunk 2]
```diff
@@ -14,6 +24,38 @@ import type { INTERNAL_Snapshot as Snapshot } from './vanilla'
 
 const { useSyncExternalStore } = useSyncExternalStoreExports
 
+// customized version of affectedToPathList
+// we need to avoid invoking getters
```
Comment (stephenh) [inline src/react.ts:28]: @dai-shi ahhh, okay, this change (skipping getters in the output of `affectedToPathList`) is why I've been confused.

I've been watching the output of `useAffectedDebugValue` (maybe a little too religiously in retrospect) and was very confused by the lack of getters.

This made me think "valtio is not working" / "doesn't know my getters were invoked", when obviously I could tell it did, b/c the re-renders still worked correctly.

I finally reproduced the behavior in `proxy-compare` and "get it"--getters (and method invocations) are always in the affected map, and are always handled correctly by `isChanged`, it is merely `affectedToPathList` that is leaving them off.

What was the rationale here? I'm sure it makes sense, just curious b/c it tripped me up. Wondering if I can work the rationale/behavior in a doc/"gotchas" update.

Thanks!

Comment (dai-shi) [reply src/react.ts:28]: Without it, it loops infinitely and hangs, AFAIR.

**Pull request merged**